subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaExecToolchainIntegrationTest.groovy

```diff
@@ -83,68 +51,208 @@ class JavaExecToolchainIntegrationTest extends AbstractIntegrationSpec {
+        "current JVM"   | "when toolchains are not configured" | false    | false
+        "executable"    | "when configured"                    | false    | true
+        "assigned tool" | "when configured"                    | true     | false
+    }
+
+    private TestFile configureProjectWithApplicationPlugin(JavaVersion compileWithVersion) {
+        buildFile << """
+            apply plugin: "application"
+
+            compileJava {
+                javaCompiler = javaToolchains.compilerFor {
+                    languageVersion = JavaLanguageVersion.of(${compileWithVersion.majorVersion})
                 }
             }
 
             application {
-                mainClass = 'App'
+                mainClass = "App"
             }
         """
+    }
 
-        file('src/main/java/App.java') << testApp()
+    private TestFile configureProjectWithoutApplicationPlugin() {
+        buildFile << """
+            plugins {
+                id 'jvm-toolchains'
+            }
 
-        when:
-        result = executer
-            .withArgument("-Porg.gradle.java.installations.paths=" + someJdk.javaHome.absolutePath)
-            .withArgument("--info")
-            .withTasks("run")
-            .run()
+            // Just outputting the JVM version that runs, instead of configuring a custom JavaCompile task
+            task run(type: JavaExec) {
+                setJvmArgs(['-version'])
+                mainClass = 'None'
+            }
+        """
+    }
 
-        then:
-        outputContains("App running with ${someJdk.javaHome.absolutePath}")
-        noExceptionThrown()
+    private TestFile configureJavaExtension(Jvm jdk) {
+        buildFile << """
+            java {
+                toolchain {
+                    languageVersion = JavaLanguageVersion.of(${jdk.javaVersion.majorVersion})
+                }
+            }
+        """
     }
 
-    private static String testApp() {
-        return """
-            public class App {
-               public static void main(String[] args) {
-                 System.out.println("App running with " + System.getProperty("java.home"));
-               }
+    private TestFile configureExecutable(Jvm jdk) {
+        buildFile << """
+            run {
+                executable = "${TextUtil.normaliseFileSeparators(jdk.javaExecutable.absolutePath)}"
+            }
+        """
+    }
+
+    private TestFile configureLauncher(Jvm jdk) {
+        buildFile << """
+            run {
+                javaLauncher = javaToolchains.launcherFor {
+                    languageVersion = JavaLanguageVersion.of(${jdk.javaVersion.majorVersion})
+                }
             }
-        """.stripIndent()
+        """
     }
 
+    private withInstallations(Jvm... jvm) {
+        def installationPaths = jvm.collect { it.javaHome.absolutePath }.join(",")
+        executer
+            .withArgument("-Porg.gradle.java.installations.paths=" + installationPaths)
+        this
```

Comment:
The main advantage of having the `withInstallations` call as a separate entity is that it can be combined with multiple "task runner actions" such as `run` and `runAndFail`, which is required for some tests.

Additionally, we already use a similar pattern with `withBuildCache()` method on the `AbstractIntegrationSpec`